src/fal/planner/plan.py

```diff
@@ -68,13 +69,25 @@ def _reduce_subgraph(
         graph.remove_node(node)
 
 
-def plan_graph(graph: nx.DiGraph) -> nx.DiGraph:
-    # Implementation of Gorkem's Critical Nodes Algorithm
-    # with a few modifications.
+def plan_graph(
+    graph: nx.DiGraph,
+    execution_plan: ExecutionPlan,
+    enable_chunking: bool = True,
+) -> nx.DiGraph:
```

Comment:
so we have two notions of `plan` getting mixed here, we have the old `ExecutionPlan` class and this method here.. curious to see what you think but should we move this method inside the `ExecutionPlan` plan class?

Reply:
I think `ExecutionPlan` and parallelization planner are two different components. The planner in my mind represents all the graph processing logic, this used to be only chunking but now it also includes applying selection/exclusion filters. On the other hand, `ExecutionPlan` is a more raw component that does argument parsing & filtering. Don't know if it should belong there.

Reply:
alright I think we are on the same page. but removing and adding nodes (via selectors) can still be ExecutionPlan 's job.. 

and planner/plan.py 's job will be chunking, and today there is not chunking so the only thing that happens is removing or adding new nodes depending on what we have decided in the `ExecutionPlan`.  I think we can move that back to `ExecutionPlan` and that class can return a readly available graph to the planner.. 

how about we create a state machine for the stages, the graph goes through and let the type system guide the way.. 

OriginGraph (today called NodeGraph)  -> FilteredGraph (output of ExecutionPlan) -> PlannedGraph (output of plan_graph method)

```
@dataclass
class GraphState: 
    graph: nx.DiGraph
```

`planner/plan.py` will look like:
```
class PlannedGraph(GraphState):
  @classmethod
   def create_planned_graph(
       filtered_graph: FilteredGraph,
        enable_chunking: bool = True,
   ) -> PlannedGraph:
    new_graph = filtered_graph.copy()
    if enable_chunking:
        # Implementation of Gorkem's Critical Nodes Algorithm
        # with a few modifications.
        for nodes in _find_subgraphs(new_graph):
            _reduce_subgraph(new_graph, nodes)

    return PlannedGraph(new_graph)
```
similarly FilteredGraph (now ExecutionPlan could have a class method that takes in an `OriginGraph` and creates an ExecutionPlan)

Reply:
If the order is static, then would it be an option to call `filter_graph()` as the first action of the `plan_graph`. Like this:

```py
def filter_graph(graph: nx.DiGraph, execution_plan: ExecutionPlan) -> nx.DiGraph:
    ...

def plan_graph(graph: nx.DiGraph, execution_plan: ExecutionPlan) -> nx.DiGraph:
    graph = filter_graph(graph, execution_plan)
    graph = optimize_graph(graph)
    return graph
```

This completely eliminates the need from client-side to know which order they should call (this is also quite similar to the existing method, where we only need to split `plan_graph` into two parts).

Reply:
yes true for the current implementation but doesn't stop future misuse.. 

this would have been a way to dictate state transitions only happen between specific types of `nx.DiGraph`.. doesn't improve our system that much right now so i am ok keeping `nx.DiGraph`s around..

Reply:
I've gave the implementation a shot at https://github.com/fal-ai/fal/pull/423/commits/7fc6702d67480d7e0ff8d6636a4d875164a4f618.